### DISCUSSION TITLE
Render engine references

### DISCUSSION THREAD

**Discussion by halx99** [axmolengine/axmol#857] [Ideas]
* https://github.com/google/filament
* https://github.com/bkaradzic/bgfx
* https://github.com/OGRECave/ogre-next
* https://github.com/DiligentGraphics/DiligentEngine - Apple Metal backend require commercial license, can free support with [MoltenVK](https://github.com/KhronosGroup/MoltenVK) or [gfx-portability](https://github.com/gfx-rs/portability)
* https://github.com/floooh/sokol
* https://github.com/LukasBanana/LLGL

**Comment by halx99** [comment]
References:
https://blog.csdn.net/android_wql/article/details/85044252

**Comment by RustyMoyher** [comment]
I'd _love_ to see this feature! There is a WIP port of cocos v4 + bgfx [here](https://github.com/Xrysnow/cocos2d-x-bgfx). Might be useful.
https://github.com/Xrysnow/cocos2d-x-bgfx

Also, there is a fork of cocos2d-x-lite w/ bgfx here:
https://github.com/quicklyslow/cocos2d-bgfx

**Comment by halx99** [comment]
Thanks, it's really useful

**Comment by Xrysnow** [comment]
CocosCreator released a new renderer for its native engine: https://github.com/cocos-creator/engine-native/tree/v3.3.0/cocos/renderer, which is vulkan-styled and very flexible. I have tried for weeks to port it but always meet problems. Do you have any interests on it?

**Comment by halx99** [comment]
Interesting, if you port success, it's really good for adxe

**Comment by halx99** [comment]
You can put your problem or your port branch at here

**Comment by Xrysnow** [comment]
Still need some days to put it in order. I will make comment here after finished.

**Comment by halx99** [comment]
Great.

Get Outlook for Android<https://aka.ms/AAb9ysg>

________________________________

**Comment by Xrysnow** [comment]
I have encountered endless memory and graphic problems when porting CocosCreator. I think the right way is to port all related js code and modify many classes, which is a huge work. I'm not sure if it worth so I put it aside :(

**Comment by aismann** [comment]
maybe this #439 is also a way for Android (but seems to be no multiplatform support)

**Comment by aismann** [comment]
I have encountered endless memory and graphic problems when porting CocosCreator. I think the right way is to port all related js code and modify many classes, which is a huge work. I'm not sure if it worth so I put it aside :(
Originally posted by @Xrysnow in #67 (comment)

@Xrysnow 
Have you a github reposity of this?
I tried tio compile it but there are some files missing

**Comment by aismann** [comment]
Update: 3.3.1 => can compile (3.4.0 not)
engine-native-3.3.1\engine-native-3.3.1\build\Debug\cocos2d.lib

**Comment by aismann** [comment]
Maybe ralylib 4.0 is also a candidate as render engine plus some more.
https://www.raylib.com/
https://github.com/raysan5/raylib/
https://github.com/raysan5/raylib/releases

**Reply by paulocoutinhox** [reply]
Raylib don't have Metal support (ios, tvos)

**Comment by Xrysnow** [comment]
@halx99 @aismann Finally I ported [renderer of cocos-engine 3.5.0](https://github.com/cocos/cocos-engine/tree/v3.5.0/native/cocos/renderer) and make it work at [this repo](https://github.com/Xrysnow/cocos2d-x/tree/gfx). 3.5.0 seems more robust than previous version to make it possible. However, it need more test and there are still many todos in it.
The ported backend is in `cocos/renderer/backend/gfx` and dependencies is in [this repo](https://github.com/Xrysnow/cocos2d-x-3rd-party-libs-bin/tree/gfx). To build it, `CC_USE_GFX` and `USE_JOB_SYSTEM_TASKFLOW` should be set in cmake. It also requires PVRVFrame as GLES provider (ANGLE seems work bad).

**Comment by aismann** [comment]
Interessting fork:
This is a renderer for [cocos2d-x](https://github.com/cocos2d/cocos2d-x) ported from [cocos-engine](https://github.com/cocos/cocos-engine).
https://github.com/Xrysnow/cocos2d-x-gfx

**Comment by Xrysnow** [comment]
> Interessting fork: This is a renderer for [cocos2d-x](https://github.com/cocos2d/cocos2d-x) ported from [cocos-engine](https://github.com/cocos/cocos-engine). https://github.com/Xrysnow/cocos2d-x-gfx

It's working right now but still need more test and feedback. I'm not sure what dose @halx99 think about it.

**Comment by aismann** [comment]
> > Interessting fork: This is a renderer for [cocos2d-x](https://github.com/cocos2d/cocos2d-x) ported from [cocos-engine](https://github.com/cocos/cocos-engine). https://github.com/Xrysnow/cocos2d-x-gfx
> 
> It's working right now but still need more test and feedback. I'm not sure what dose @halx99 think about it.

Maybe you can add it here on a PR?
So we all can test it.
@halx99 can make a branch for this

**Comment by halx99** [comment]
I'm not sure, maybe @Xrysnow can create PR for us, what rendering performance about the cocos-engine renderer.

**Comment by aismann** [comment]
@halx99 
maybe its better laying on discussion board?

**Comment by aismann** [comment]
https://skia.org/

Skia is an open source 2D graphics library which provides common APIs that work across a variety of hardware and software platforms. It serves as the graphics engine for Google Chrome and ChromeOS, Android, Flutter, and many other products.

**Comment by paulocoutinhox** [comment]
Will be nice if we have BGFX (https://github.com/bkaradzic/bgfx) that works on:

**Supported rendering backends:**

1. Direct3D 9
2. Direct3D 11
3. Direct3D 12
4. GNM (only for licensed PS4 developers, search DevNet forums for source)
5. Metal
6. OpenGL 2.1
7. OpenGL 3.1+
8. OpenGL ES 2
9. OpenGL ES 3.1
10. Vulkan
11. WebGL 1.0
12. WebGL 2.0
13. WebGPU/Dawn (experimental)

**Supported platforms:**

1. Android (14+, ARM, x86, MIPS)
2. FreeBSD
3. iOS (iPhone, iPad, AppleTV)
4. Linux
5. MIPS Creator CI20
6. OSX (10.12+)
7. PlayStation 4
8. RaspberryPi
9. UWP (Universal Windows, Xbox One)
10. Wasm/Emscripten
11. Windows (XP, Vista, 7, 8, 10)

**Others**

It is already well tested in other engines, real games for pc like minecraft and mobile games on store.

On their README have all informations.

**Comment by halx99** [comment]
`bgfx` does not fit well with Axmol’s renderer framework, so I decided to develop a custom backend (RHI). Since Oct, 2025, the Direct3D11 implementation has been completed. Axmol now supports OpenGL, Metal, and Direct3D11, with Vulkan planned for future integration.

**Reply by aismann** [reply]
Maybe that's a start?
https://github.com/PatriceJiang/cocos2d-x-lite/tree/ohos/cocos/renderer/gfx-vulkan

**Reply by paulocoutinhox** [reply]
What exactly is the problem? Can't this be circumvented and resolved? I believe the gains will be significant, and you can dedicate more time to engine issues and version 3.

**Reply by halx99** [reply]
axmol v3 now have own modern RHIs: gl, d3d11, d3d12, vulkan, metal

**Reply by paulocoutinhox** [reply]
Wow. Thanks. Need study a lot to understand how to render on each engine. Congratz.
